### PR TITLE
Fix mongodb lookup

### PR DESCRIPTION
**Pull request by marcosdiez**
This PR has moved to https://github.com/ansible-collections/mongodb/pull/38
--------------------

##### SUMMARY
I am the original author of the plugin. 
After receiving an email with a support question, I realized:
  * the documentation has been update (by somebody else) and had an invalid ansible syntax, therefore it was not working
  * the plugin was not loudly complaining in case pymongo was not present
  * the plugin was not properly displaying error messages in ansible 2.8 ( when I wrote it, ansible was at version 2.3)

##### ISSUE TYPE
- Bugfix Pull Request
- Docs Pull Request

##### COMPONENT NAME

mongodb_lookup

##### ANSIBLE VERSION

```
ansible 2.8.0.dev0 (fix_mongodb_lookup 8f744587e9) last updated 2018/10/06 13:52:09 (GMT -200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/mdiez/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /mnt/extra/u/rootfs/home/ubuntu/code/ansible/lib/ansible
  executable location = /mnt/extra/u/rootfs/home/ubuntu/code/ansible/bin/ansible
  python version = 3.5.2 (default, Nov 23 2017, 16:37:01) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

The biggest problem that I fixed was in the example usage in the documentation:

As one can see here: https://docs.ansible.com/ansible/latest/plugins/lookup/mongodb.html

it eventually says
```
    - debug: msg="Mongo has already started with the following PID [{{ item.pid }}]"
      with_mongodb: "{{mongodb_parameters}}"

```

which makes no sense since the mongodb lookup does not return an `item` object.

The second change is that sometime between `ansible 2.3` and `devel`, my plugin stopped writing error messages. So I am now using `AnsibleError`, which works !

Last but not least somehow I forgot to check if `pymongo` was installed. And it took me 3 years to find it out. Oops...

I also added integration tests to make sure things won't break in the future!

### CODE REVIEW

**Comment by marcosdiez** [comment]
@Lujeni @Dufgui hi guys! Two years ago you reviewed my ansible mongodb lookup module. It has some issues and here is a PR with some code fixes. Would you care to code review them ?

Thanks!

**Review by Dufgui — APPROVED** [review]
Ok for me.

File: lib/ansible/plugins/lookup/mongodb.py
Comment (Lujeni) [inline lib/ansible/plugins/lookup/mongodb.py]: not a big deal, you can add a space :)

File: lib/ansible/plugins/lookup/mongodb.py
Comment (Lujeni) [inline lib/ansible/plugins/lookup/mongodb.py]: you should use `loop`, `with_items` is deprecated :)

File: lib/ansible/plugins/lookup/mongodb.py
Comment (Lujeni) [inline lib/ansible/plugins/lookup/mongodb.py]: imho, you can keep your message KISS and more straighforward: 
e.g. "pymongo library is required for this module"

**Review by Lujeni — CHANGES_REQUESTED** [review]
Thanks for the PR.
btw Can you rebase / squash your commit pls ?

Ciao

File: lib/ansible/plugins/lookup/mongodb.py
Comment (marcosdiez) [reply lib/ansible/plugins/lookup/mongodb.py]: I did not understand, sorry. Could you please explain it with other words ?

File: lib/ansible/plugins/lookup/mongodb.py
Comment (marcosdiez) [reply lib/ansible/plugins/lookup/mongodb.py]: I would prefer not to, because this is a lookup module and therefore it must be in master node instead of the target node. Your message may confuse users who are not experts in ansible.

**Comment by marcosdiez** [comment]
Hello @Lujeni !

I did some of the modifications you requested, added 5 more examples and some argumentation about other documentation. 
Could you please check now ?

**Comment by mattclay** [comment]
bot_status

File: lib/ansible/plugins/lookup/mongodb.py
Comment (Lujeni) [reply lib/ansible/plugins/lookup/mongodb.py]: blank line after your import

File: lib/ansible/plugins/lookup/mongodb.py
Comment (Lujeni) [reply lib/ansible/plugins/lookup/mongodb.py]: in this case, you should use `primary` instead of `master`

Comment (marcosdiez) [reply lib/ansible/plugins/lookup/mongodb.py]: understood.! fixed!

**Comment by marcosdiez** [comment]
@Lujeni please please please please don't forget me !

**Comment by marcosdiez** [comment]
@Lujeni please please please please please please please please  don't forget me !

**Comment by marcosdiez** [comment]
@Lujeni @Lujeni @Lujeni @Lujeni @Lujeni @Lujeni  don't forget me !

**Comment by marcosdiez** [comment]
@Lujeni  I even added integration tests.... all I need is an OK from you. Please...

**Comment by marcosdiez** [comment]
@Lujeni wheeeeeeeeeeeeeeeeere aaaaaaaaaaaaaareeeeeeeeeee youuuuuuuuuuuuuuuu?

**Comment by marcosdiez** [comment]
@Lujeni please code review my PR

**Comment by mdiez-modus** [comment]
@Lujeni have you forgotten me?

**Comment by marcosdiez** [comment]
@beak42py could you please code review ?

**Review by beak42py — APPROVED** [review]

**Comment by marcosdiez** [comment]
bot_status

**Comment by marcosdiez** [comment]
@mattclay hi! I have two approvals on GitHub. What is now needed for this PR to be merged ?

**Comment by marcosdiez** [comment]
@dagwieers is there anything I can do to help the PR get merged?

**Comment by dagwieers** [comment]
We can start a MongoDB Working Group to get mongodb developers review and approve PRs. This could help moving things forward. I already tagged all MongoDB related issues and PRs, so it would only be a small step

**Comment by marcosdiez** [comment]
@dagwieers it has been another 11 days and... nothing ? Can I help you getting this merged ? I have users who are counting on this bugfix

**Comment by dagwieers** [comment]
@marcosdiez I can review, but I cannot test. And since I have no mongodb background I can't evaluate properly. So it is better to set up a group of peers to collaborate. Let me set up the MongoDB Working Group and then we can ask the group.

File: lib/ansible/plugins/lookup/mongodb.py
Comment (dagwieers) [reply lib/ansible/plugins/lookup/mongodb.py]: These imports should go after the DOCUMENTATION, EXAMPLES and RETURN strings.

File: lib/ansible/plugins/lookup/mongodb.py
Comment (dagwieers) [inline lib/ansible/plugins/lookup/mongodb.py]: Please split up the description items. The first item should be an explanation of the description.
```suggestion
                - Sorting rules.
                - Please use the strings C(ASCENDING) and C(DESCENDING) to set the order.information.
```

**Review by dagwieers — DISMISSED** [review]
Only a few remarks. I cannot test, but the integration tests are promising.

**Comment by dagwieers** [comment]
@marcosdiez 2 approvals have no meaning to @ansibot. What you need are 2 community shipits from maintainers. And that is not possible currently. Only if we have a MongoDB Working Group we can give members shipit rights.

File: lib/ansible/plugins/lookup/mongodb.py
Comment (gundalow) [reply lib/ansible/plugins/lookup/mongodb.py]: `control node` = Where ansible-playbook is run
`remote node` = system to configure

**Comment by gundalow** [comment]
@marcosdiez Looks like their are just a few minor things to address, could you take a look?

We are very close to the Ansible 2.8 cut off https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_8.rst

**Comment by marcosdiez** [comment]
Hi John,

I'll check it tonight after my working hours.

Thanks for your time!

Marcos

**Comment by mdiez-modus** [comment]
@gundalow changes made :)

**Comment by sivel** [comment]
@marcosdiez @mdiez-modus I don't see that any new commits have been pushed. I see you have commented from an account different than the one associated with this PR.  Did you push to the correct remote?

**Comment by mdiez-modus** [comment]
@sivel I am not a very bright man... please check again :)

**Comment by mdiez-modus** [comment]
All done! Rebased and all tests passed!

**Comment by marcosdiez** [comment]
@dagwieers I made all the changes your requested.
@gundalow @sivel what else is need to have this merged ? The PR was actually done in October and it's a bugfix to something that I wrote myself. Should it really take that long?

File: test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml
Comment (mattclay) [inline test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:20]: Can the connection timeout be lowered? This task takes 30 seconds to complete:

```
13:42 TASK [lookup_mongodb : assert we complain properly if we can't connect to mongo] ***
14:12 There was an exception on the mongodb_lookup: unable to connect to database: localhost:27017: [Errno 111] Connection refused
```

File: test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml
Comment (mattclay) [inline test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:4]: The test should not assume that `pymongo` is not already installed.

File: test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml
Comment (mattclay) [inline test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:36]: What issue(s) did you encounter when trying to use the `apt` module?

File: test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml
Comment (marcosdiez) [reply test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:20]: Agreed. I changed it to 5 seconds in the test.

**Comment by marcosdiez** [comment]
Hey, what else needs to be done here ?
I am the original author of the module.
I found a bug on it, and I sent a PR to fix it. 
I also added passing integration tests to it.

I implemented every suggestion give by @Lujeni @gundalow and @mattclay 

what else needs to be done for this to be merged?

**Comment by marcosdiez** [comment]
@dagwieers I beleive I did everything that you requested. Could you review and approve my PR ? It's contains only bugfixes and examples, no new features.

**Comment by marcosdiez** [comment]
@dagwieers please don't forget me!

File: test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml
Comment (gundalow) [reply test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:36]: @marcosdiez Could we please get an answer to this?

File: test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml
Comment (gundalow) [reply test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:4]: This doesn't look like it's been resolved

Comment (marcosdiez) [reply test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:4]: I believe it is. I check if "pymongo is required" exactly 5 lines below this one.

File: test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml
Comment (marcosdiez) [reply test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:36]: That's a hard one. I don't remember. I wrote this code in October. If this is blocking the PR from being merged, I can try to reproduce the issue. In my humble opinion this is secondary, since 
- this is only a test and not the main code itself
- yes, it could be better but it actually works

**Comment by marcosdiez** [comment]
@dagwieers , @gundalow I still don't know why this PR has not been merged...

**Comment by marcosdiez** [comment]
@dagwieers , @gundalow I still don't know why this PR has not been merged...

**Comment by marcosdiez** [comment]
@dagwieers , @gundalow I still don't know why this PR has not been merged...

**Comment by marcosdiez** [comment]
@dagwieers , @gundalow I still don't know why this PR has not been merged...

**Comment by gundalow** [comment]
bot_status

File: test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml
Comment (gundalow) [reply test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:36]: Agreed, I don't think this blocks

File: test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml
Comment (gundalow) [reply test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:4]: I think the problem is that `pymongo` may already be installed

**Comment by marcosdiez** [comment]
@rhysmeister I believe you are the person in charge for everything MongoDB related in Ansible.
I have this PR that has been opened since 2018 and all it needs is a code review and an approval.
Could you please take a look ? Thank you!

**Comment by rhysmeister** [comment]
Hello @marcosdiez actually I'm not. I've just submitted a few PRs. I don't have any authority I'm afraid. 

There has been some chatter about moving the mongodb modules to a community support collection. Perhaps you'd like this included?

**Comment by marcosdiez** [comment]
Hy Rhys!

Honestly, all I want is my 2 year old PR merged.

If you could code review it, then it would be easier to get the merge approved.

If not, well, please include it in the mongodb collection.

If you know anything I can do to get it approved please let me know.

Thanks

Marcos

Am 5. Februar 2020 21:40:28 MEZ schrieb Rhys Campbell <notifications@github.com>:
>Hello @marcosdiez actually I'm not. I've just submitted a few PRs. I
>don't have any authority I'm afraid. 
>
>There has been some chatter about moving the mongodb modules to a
>community support collection. Perhaps you'd like this included?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/ansible/ansible/pull/46579#issuecomment-582601811

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by rhysmeister** [comment]
Hello Marcos,

I've tested your plugin... https://github.com/rhysmeister/TestMongoDBLookup

Looks good although a couple of comments. Error given when pymongo installed...

> fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'mongodb'. Error was a <class 'ansible.errors.AnsibleError'>, original message: pymongo is required in the master node (this machine) for mongodb lookup."}
 
Probably better without the "unhandled exception" stuff?

I also got the following exception...

> fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'mongodb'. Error was a <class 'KeyError'>, original message: 'database'"}

I realise that's because I was missing a few mandatory parameters but I think the message could be nicer/ better handled.

That's about all I can do at the moment. Good luck!

Rhys

**Comment by marcosdiez** [comment]
Hey, @rhysmeister ,
thanks for testing my PR.

But following your instructions, you did not test the code from the PR, just the normal plugin in ansible.

I believe this is how it should be tested:

```
docker run -tid --rm -p 27017:27017 --name mongodb rhyscampbell/dockermongodb

virtualenv venv # or python3 -m venv venv
source venv/bin/activate
git clone --depth 1 --single-branch --branch fix_mongodb_lookup git@github.com:marcosdiez/ansible.git
source ansible/hacking/env-setup
pip install jinja2 
pip install pyyaml
ansible-playbook -l localhost test_mongodb.yml
```

the result is:

fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'mongodb'. Error was a <class 'ansible.errors.AnsibleError'>, original message: pymongo is required in the control node (this machine) for mongodb lookup."}

after `pip instsall pymongo` ansible runs successfully.

**Comment by rhysmeister** [comment]
Hello @marcosdiez 

Your plugin is included in the lookup_plugins directory. Ansible will pick this up rather than the default (old) one. You can see it does because it creates a python cache dir...

ls -lh lookup_plugins/
total 24
drwxr-xr-x  3 rhyscampbell  staff    96B 14 Feb 12:02 __pycache__
-rw-r--r--  1 rhyscampbell  staff   9,6K 14 Feb 12:03 mongodb.py

You can add a syntax error to the file if you're not convinced. The playbook then breaks, i.e.

> An exception occurred during task execution. To see the full traceback, use -vvv. The error was: NameError: name 'sdfsdfgsdagfsdfg' is not defined
> fatal: [localhost]: FAILED! => {"msg": "Unexpected failure during module execution."}

Rhys

**Comment by marcosdiez** [comment]
Sorry, @rhysmeister, you are right.
I tried again and this time it happend exactly as you mentioned.

**Comment by marcosdiez** [comment]
Hey @klingac

I have  PR to fix issues in Ansible's MongoDB lookup that is opened for more than 3 years. All I need is somebody like you to approve it so it can be merged.

Could you please take a look at it?

Thank you
Marcos

File: test/integration/targets/lookup_mongodb/tasks/main.yml
Comment (klingac) [inline test/integration/targets/lookup_mongodb/tasks/main.yml]: Is check for Xenial really necessery? We have at least newer LTS version for Ubuntu. I just tried that with 18.04 and it works.
```suggestion
  when: ansible_distribution == 'Ubuntu'
```

File: lib/ansible/plugins/lookup/mongodb.py
Comment (klingac) [inline lib/ansible/plugins/lookup/mongodb.py:69]: Plase also add extra_parameters. You are mentioning it in examples, but not in documentation. It could be little bit confusing
```suggestion
            default: "[]"
        extra_connection_parameters:
            description:
                - Extra connection parameters, eg. { "readPreference": "nearest", "ssl" : True , "ssl_certfile": "/etc/self_signed_certificate.pem" }
            type: 'dict'
            default: '{}'
```

**Review by klingac — COMMENTED** [review]
Hello Marcos,

I've just tested your plugin on Ubuntu 18.04 LTS. Everything seems to be OK for me but I have two suggestions - see below. Those suggestion are only about tests and documentation. Plugin seems to be functioning well.

File: test/integration/targets/lookup_mongodb/tasks/main.yml
Comment (mdiez-modus) [reply test/integration/targets/lookup_mongodb/tasks/main.yml]: No. It's not and it has been removed. When this was done, bionic (18.04) did not exist, xenial (16.04) was new and there were still ansible tests for 14.04, which either did not contain mongodb or pymongo, which is why this lines was there. Thanks for noticing.

Comment (marcosdiez) [reply test/integration/targets/lookup_mongodb/tasks/main.yml]: No. It's not and it has been removed. When this was done, bionic (18.04) did not exist, xenial (16.04) was new and there were still ansible tests for 14.04, which either did not contain mongodb or pymongo, which is why this lines was there. Thanks for noticing.

File: lib/ansible/plugins/lookup/mongodb.py
Comment (marcosdiez) [reply lib/ansible/plugins/lookup/mongodb.py:69]: Once again, thanks for having noticed this. It's now added into the documentation as well.

**Comment by marcosdiez** [comment]
@klingac I did the two changes you requested and the tests are passing. Do you need anything ? If not, could you please approve this PR so it can get merged ? Thanks!

**Review by klingac — APPROVED** [review]
Looks OK to me.
Please merge this PR

**Comment by marcosdiez** [comment]
@dagwieers I just got an approval from @klingac . Is this PR now eliable to be merged ? If not, what else is needed?

**Comment by marcosdiez** [comment]
@dagwieers Please take a look at this PR!

**Comment by gundalow** [comment]
rebuild_merge

**Comment by gundalow** [comment]
FYI mongodb will soon have its own collection. For those wishing to join the mongodb working group you way wish to watch (subscribe) to https://github.com/ansible-collections/mongodb

The code that's currently in Ansible/Ansible will be moved to this new location within the next he next month.

See https://github.com/ansible-collections/overview for more info.

File: test/integration/targets/lookup_mongodb/aliases
Comment (mattclay) [inline test/integration/targets/lookup_mongodb/aliases:6]: Add one more entry here:

```
skip/python2.6  # filters are controller only, and we no longer support Python 2.6 on the controller
```

**Comment by mkrizek** [comment]
!waffling needs_revision

**Comment by marcosdiez** [comment]
So... I pass all tests, what's next ?

**Comment by marcosdiez** [comment]
Since this is a PR to fix a bug and I am trying to merge it for the last 2 years, could we just merge it against 2.9.x as a bugfix ?

**Comment by sivel** [comment]
This PR would need to be resubmitted to https://github.com/ansible-collections/mongodb to be considered for merge.

All mongodb functionality has been removed from this repository and migrated to the above repo.

**Comment by marcosdiez** [comment]
This PR has been moved here https://github.com/ansible-collections/mongodb/pull/38 and has been merged!

**Pull request closed without merging**
